### PR TITLE
Fix SSA redundant CFG computation

### PR DESCRIPTION
**Pull request by sklam**
Addresses SSA reconstruction performance problem seen in #5611.

### CODE REVIEW

**Comment by stuartarchibald** [comment]
This patch looks good, the CFG and derived properties like dominance frontiers are SSA transform invariant so compute once and cache makes sense.

File: numba/core/ssa.py [hunk 1]
```diff
@@ -60,19 +62,19 @@ def _run_ssa(blocks):
         _logger.debug("Replaced assignments: %s", pformat(defmap))
         # Fix up the RHS
         # Re-associate the variable uses with the reaching definition
-        blocks = _fix_ssa_vars(blocks, varname, defmap)
+        blocks = _fix_ssa_vars(blocks, varname, defmap, cfg, df_plus)
```
Comment (stuartarchibald) [inline numba/core/ssa.py:65]: Is it worth asserting that the CFG hasn't changed on exit? CFGs are comparable with `__eq__`

**Comment by stuartarchibald** [comment]
One minor thing to resolve else looks good, thanks for the patch.

File: numba/core/ssa.py (hunk 1 above)
Comment (sklam) [reply numba/core/ssa.py:65]: That would require recomputing the cfg and would be too expensive given that SSA is already too costly in our compilation pipeline.  We have to rely on making sure the SSA algorithm are not mutating the CFG.

File: numba/core/ssa.py [hunk 2]
```diff
@@ -63,6 +63,12 @@ def _run_ssa(blocks):
         # Fix up the RHS
         # Re-associate the variable uses with the reaching definition
         blocks = _fix_ssa_vars(blocks, varname, defmap, cfg, df_plus)
+
+    # Post-condition checks.
+    # CFG invariant
+    cfg_post = compute_cfg_from_blocks(blocks)
+    if cfg_post != cfg:
+        raise AssertionError("CFG mutated in SSA pass")
```
Comment (stuartarchibald) [inline numba/core/ssa.py:71]: Should this be a `errors.CompilerError` perhaps?

Comment (sklam) [reply numba/core/ssa.py:71]: yes indeed

File: numba/core/ssa.py (hunk 1 above)
Comment (stuartarchibald) [reply numba/core/ssa.py:65]: out of band, cost of a single CFG recomputation on loop exit was considered worth it

**Pull request merged**